File: 3. Behavioral Design Patterns/Strategy/Javascript/StrategyTemplate.js
// Strategy Pattern – JavaScript Template

class Strategy { execute(data){ return data } }
class ConcreteStrategyA extends Strategy { execute(data){ return data } }
class ConcreteStrategyB extends Strategy { execute(data){ return data } }

class Context {
  constructor(strategy){ this.strategy = strategy }
  setStrategy(strategy){ this.strategy = strategy }
  doWork(data){ return this.strategy.execute(data) }
}

;(function main(){ const ctx = new Context(new ConcreteStrategyA()); ctx.doWork([]) })()

File: 3. Behavioral Design Patterns/Strategy/Go/StrategyTemplate.go
package main

// Strategy Pattern – Go Template

type Strategy interface{ Execute([]int) []int }

type ConcreteStrategyA struct{}
func (ConcreteStrategyA) Execute(data []int) []int { return data }

type ConcreteStrategyB struct{}
func (ConcreteStrategyB) Execute(data []int) []int { return data }

type Context struct{ s Strategy }
func (c *Context) Set(s Strategy){ c.s = s }
func (c Context) Do(data []int) []int { return c.s.Execute(data) }

func main(){ ctx := Context{ s: ConcreteStrategyA{} }; _ = ctx.Do([]int{}) }
